main/file.c

```diff
@@ -213,52 +285,193 @@ static void on_send(sub_send_cb_data_t *send_data){
 
         if(current_time_ms - sub_struct.last_tx_time_ms > sub_struct.tx_nack_timeout_ms){
 
-            if(xSemaphoreTake(sub_struct.mutex, portMAX_DELAY) == pdTRUE){
-
-                sub_struct.interval_ms = sub_struct.transmit_periods[SLEEP_MS];
-
-                xSemaphoreGive(sub_struct.mutex);
-
-            }
+            set_interval_ms(sub_struct.transmit_periods[SLEEP_MS]);
         }
     }
 }
 
 static void on_recive(sub_recv_cb_data_t *recv_data){
 
-    uint8_t obc_mac_addr[ESP_NOW_ETH_ALEN] = OBC_MAC_ADDRESS;
+    uint8_t mcb_mac_addr[ESP_NOW_ETH_ALEN] = MCB_MAC_ADDRESS;
+
+    if(memcmp(recv_data->dest_mac, sub_struct.dev_mac_address, ESP_NOW_ETH_ALEN)!=0){
+        return;
+    }
 
-    if(memcmp(recv_data->mac, obc_mac_addr, ESP_NOW_ETH_ALEN)==0){
+    if(memcmp(recv_data->src_mac, mcb_mac_addr, ESP_NOW_ETH_ALEN)!=0){
 
-        if(recv_data->data_size == STATE_MSG_SIZE && recv_data->data[0] < ENUM_MAX){
+        return;
+    }
 
-            if(xSemaphoreTake(sub_struct.mutex, portMAX_DELAY) == pdTRUE){
+    if(recv_data->data_size == STATE_MSG_SIZE && recv_data->cmd.raw[0] < ENUM_MAX){
 
-                sub_struct.interval_ms = sub_struct.transmit_periods[recv_data->data[0]];
+        uint64_t interval = sub_struct.transmit_periods[recv_data->cmd.raw[0]];
 
-                xSemaphoreGive(sub_struct.mutex);
-            }
-        }
+        set_interval_ms(interval);
     }
     else{
-
         recv_cb_cmd_t temp_data;
 
-        memcpy(&temp_data, recv_data->data, recv_data->data_size);
+        memcpy(&temp_data, &recv_data->cmd, recv_data->data_size);
 
         if(sub_struct._on_data_rx == NULL){
 
             if(recv_data->data_size > MAX_DATA_LENGTH){
                     
-                    return;
+                return;
             }
 
-            xQueueSend(sub_struct.rx_queue, temp_data, 0);
+            xQueueSend(sub_struct.rx_queue, &temp_data, 0);
         }
         else{
             
             sub_struct._on_data_rx(temp_data.raw, recv_data->data_size);
         }
     }
-    
+}
+
+static void send_packet(){
+
+    sub_send_cb_data_t send_data;
+
+    uint8_t mcb_mac_addr[ESP_NOW_ETH_ALEN] = MCB_MAC_ADDRESS;
+
+    if(sub_struct._data_to_transmit != NULL){
+
+        sub_struct._data_to_transmit(sub_struct.tx_buffer, MAX_DATA_LENGTH, &sub_struct.tx_data_size);
+    }
+
+    if(sub_struct.tx_data_size <= MAX_DATA_LENGTH && sub_struct.tx_data_size > 0){
+        
+        esp_now_send(mcb_mac_addr, sub_struct.tx_buffer, sub_struct.tx_data_size);
+    }
+
+    if(xQueueReceive(sub_struct.send_cb_queue, &send_data, 0) == pdTRUE){
```

Comment:
Tak szybko raczej nie dostaniesz zwrotki, bo to wysłanie jest obsługiwane przez inny task. Więc najprawdopodobniej będziesz odbierał tutaj wcześniejszą informację